csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs

```diff
@@ -140,6 +140,13 @@ public IDisposableReadOnlyCollection<DisposableNamedOnnxValue> Run(IReadOnlyColl
             int inputIndex = 0;
             foreach (var input in inputs)
             {
+                // guard against accidental DisposableNamedOnnxValue inputs flowing through
+                // as the code is not equipped to handle that
+                if (input is DisposableNamedOnnxValue)
+                {
+                    throw new NotSupportedException("Input of type 'DisposableNamedOnnxValue' is not supported");
+                }
```

Comment:
Can this message tell them what they should do instead? It's probably a little complicated to explain why it's not supported, so instead telling them how to fix the issue would be good.

Reply:
Updated the error message.